File: 0155-min-stack/0155-min-stack.js
/*
1. maintain min so far at each node in stack'
2. when element is pushed,
        compare element and min of top, store min in that element
3. getMin()
        return min of top element
*/
var MinStack = function() {
    this.elements = []
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    this.elements.push({
        val: val,
        min: this.elements.length === 0 ? val : Math.min(val, this.getMin())
    })
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    this.elements.pop()
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.elements[this.elements.length - 1]?.val
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.elements[this.elements.length - 1].min
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.elements = []

    def push(self, val: int) -> None:
        self.elements.append({
            "val": val,
            "minimum": val if len(self.elements) == 0 else min(self.elements[-1]["minimum"], val)
        })

    def pop(self) -> None:
        return self.elements.pop()["val"]

    def top(self) -> int:
        return self.elements[-1]["val"]

    def getMin(self) -> int:
        return self.elements[-1]["minimum"]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()